### ISSUE TITLE
string parameters not URL-encoded

### ISSUE THREAD

**Issue by JonathonReinhart** [pyapi-gitlab/pyapi-gitlab#125]
For `getproject()` at least, the slash needs manually URL-encoded by the caller. Example:

```
g.getproject('foo/bar')      # fails
g.getproject('foo%2Fbar')    # works
```

pyapi-gitlab should URL-encode all string parameters before making the request.

**Comment by Itxaka** [comment]
Hey @JonathonReinhart , thanks for reporting this. What version of Gitlab are you using? 
I want to create a test while fixing this so I need to spin a container with the same version.

Thanks!

**Comment by JonathonReinhart** [comment]
I believe it is version 7.9 CE, but I'll have to confirm when I'm back in
the office.

**Comment by Itxaka** [comment]
Indeed! It should happen on every version as I expect only an id to be passed instead of a project name with namespace!

**Comment by JonathonReinhart** [comment]
It's a pretty convenient feature of the API. However, it's not orthogonal -
the same feature doesn't work for milestones, or example. To find a
milestone by project namespace/name and milestone title, you can use the
project namespace/name, but have to iterate over all milestones to find the
matching title.

**Comment by JonathonReinhart** [comment]
GitLab 7.9.1 is where I'm seeing this.

**Comment by Itxaka** [comment]
@JonathonReinhart Please check with the new version as this should be fixed now. Thanks for the report!

**Comment by JonathonReinhart** [comment]
Nice work! This appears to be fixed for only that one method, though.  This actually needs to happen for every method that uses `self.projects_url` and a `project_id`. This includes:
- `getproject`
- `getprojectevents`
- `deleteproject`
- `getprojectmembers`
- `addprojectmember`
- `editprojectmember`
- `deleteprojectmember`
- `getprojecthooks`
- `getprojecthook`
- `addprojecthook`
- `editprojecthook`
- `deleteprojecthook`
- `getbranches`
- `getbranch`
- `createbranch`
- `deletebranch`
- `protectbranch`
- `unprotectbranch`
- `createforkrelation`
- `removeforkrelation`
- `createfork` (maybe, haven't tried)
- `getprojectissues`
- `getprojectissue`
- `createissue`
- `editissue`
- `getmilestones` (I know this one for sure, as I tried it)
- `getmilestone` (...)
- `createmilestone`
- `editmilestone`
- `getdeploykeys`
- `getdeploykey`
- `adddeploykey`
- `deletedeploykey`
- `getmergerequests`
- `getmergerequest`
- `getmergerequestcomments`
- `createmergerequest`
- `updatemergerequest`
- `acceptmergerequest`
- `addcommenttomergerequest`
- `getsnippets`
- `getsnippet`
- `createsnippet`
- `getsnippetcontent`
- `deletesnippet`
- `getrepositories`
- `getrepositorybranch`
- `protectrepositorybranch`
- `unprotectrepositorybranch`
- `getrepositorytags`
- `createrepositorytag`
- `getrepositorycommits`
- `getrepositorycommit`
- `getrepositorycommitdiff`
- `getrepositorytree`
- `getrawfile`
- `getrawblob`
- `getcontributors`
- `compare_branches_tags_commits`
- `getfilearchive`
- `getissuewallnotes`
- `getissuewallnote`
- `createissuewallnote`
- `getsnippetwallnotes`
- `getsnippetwallnote`
- `createsnippetewallnote`
- `getmergerequestwallnotes`
- `getmergerequestwallnote`
- `createmergerequestewallnote`
- `createfile`
- `updatefile`
- `getfile`
- `deletefile`
- `setgitlabciservice`
- `deletegitlabciservice`
- `getlabels`
- `createlabel`
- `deletelabel`
- `editlabel`

**Comment by JonathonReinhart** [comment]
The way it is right now, `getproject` behaves differently than all of the others.

**Comment by Itxaka** [comment]
Indeed you are rigth! Reopening

**Comment by bsmr** [comment]
Just one more thing as I just have some troubles because of this: `.` (dots) in project names should be escaped with `%2E`. A project name (example _without_ the _namespace_) like `foo.bar.example` can not be found without the above replacement (e.g. `foo%2Ebar%2Eexample` will be found).

Have a look at [Dot in username or project name makes Project not found via API](https://github.com/gitlabhq/gitlabhq/issues/8553), please.

**Comment by Sudo-Kid** [comment]
First of many PR's to come https://github.com/pyapi-gitlab/pyapi-gitlab/pull/226.
